draft/XXXX-upload-system/README.md

```diff
@@ -0,0 +1,126 @@
+PUT /api/upload/nested_json/{workspace}/{graph}
+PUT /api/upload/newick/{workspace}/{graph}
+
+Body: A chunk of the original data
+
+Query Parameters: {
+  upload: id (returned by the first call),
+}
+```
+
+The `upload` query parameter is part of the information returned by the first call, and is used to denote the upload object to which the sent chunk of data belongs to. Since an upload is tied to a specific user, this endpoint is gated behind user login. Furthermore, a `PUT` request (by a logged in user) to an upload endpoint targeting an upload that does not belong to this user, should return `403 Forbidden`.
+
+
+New endpoints for getting upload information are also required. These would be of the following formats:
+
+```
+GET /api/upload/uploads
+```
+
+This endpoint is tied to the currently authenticated user, and returns all of the uploads which belong to the user. If the need for a global upload listing endpoint arises, this may have to be revised.
+
+
+### Intermediate Processing of Chunks
+
+Each data type uploader (csv, newick, etc.) will be responsible for specific logic regarding how chunked uploads are handled and stored in the "hidden" arangodb collection. Given a chunk of data, the uploader needs to be able to process/keep track the following information:
+
+* How many rows/documents are fully contained in that chunk
+* How to convert/transform a row/document from its original format, into the format required for arangodb storage
+* The current position or "context" in the full document (in graphs for example, are we currently looking at nodes?, links?, etc.)
+
+More than likely, a chunk will not often start/end exactly on the boundary between rows/documents. In this case, any "incomplete" rows/documents will need to stored in a specific location (perhaps on the upload document) until the data which completes that record arrives. Storing this data in memory is not an option, as compatibility across multiple workers must be ensured.
+
+The approach above assumes that these chunks of data are being sent **in order**. If there is a need for sending these chunks out of order, more logic would be required to keep track of "incomplete" rows/documents, and ensuring that in the future, they are able to be completed properly.
+
+#### Alternate Approach
+Another way to handle chunked uploads would be to explicitly split chunks along the boundaries of rows/documents, instead of doing this on arbitrary bytes. This would reduce the logic of chunk processing for each uploader. However, this approach would add increased complexity to, as well as place the burden of data processing on, the client application. Specific processing for each data type would be required, in order to determine the boundaries for each row/document. Furthermore, these rows/documents may be (and in practice often are) individually small, which could increase the overhead of the upload system, decreasing the overall upload speed.
```

Comment:
Ah this is what I was describing above, so perhaps I did not fully understand the mainline approach.